In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import KFold , cross_val_score , train_test_split ,GridSearchCV
from sklearn.neighbors import KNeighborsClassifier

In [2]:
glass = pd.read_csv("Glass.csv")

In [3]:
glass.head()

RI     Na    Mg    Al     Si     K    Ca   Ba   Fe  Type
0  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.0  0.0     1
1  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.0  0.0     1
2  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.0  0.0     1
3  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.0  0.0     1
4  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.0  0.0     1

In [5]:
glass['Type'].value_counts() #  getting different types of animals

2    76
1    70
7    29
3    17
5    13
6     9
Name: Type, dtype: int64

In [6]:
# splitting  zoo dataset into x and y 

x = glass.iloc[:,1:-1]
y = glass.iloc[:,-1]

In [7]:
x.head()

Na    Mg    Al     Si     K    Ca   Ba   Fe
0  13.64  4.49  1.10  71.78  0.06  8.75  0.0  0.0
1  13.89  3.60  1.36  72.73  0.48  7.83  0.0  0.0
2  13.53  3.55  1.54  72.99  0.39  7.78  0.0  0.0
3  13.21  3.69  1.29  72.61  0.57  8.22  0.0  0.0
4  13.27  3.62  1.24  73.08  0.55  8.07  0.0  0.0

In [8]:
y.head()

0    1
1    1
2    1
3    1
4    1
Name: Type, dtype: int64

In [11]:
#using train test split on x and y
x_train,x_test,y_train,y_test = train_test_split(x,y, test_size = 0.20, random_state = 42 )

### Fitting KNN model

In [12]:
model = KNeighborsClassifier(n_neighbors=10)
knn = model.fit(x_train, y_train)
results = cross_val_score(model, x_train, y_train, cv=2)

In [13]:
results

array([0.61627907, 0.61176471])

### predicting x test data

In [14]:
y_pred =model.predict(x_test)

In [15]:
results  = pd.DataFrame({'actual':y_test,'Predicted':y_pred})

In [13]:
results

actual  Predicted
84       1          1
55       1          1
66       1          4
67       1          1
45       1          1
39       6          6
22       1          1
44       1          1
10       1          1
0        1          1
18       4          4
30       6          6
97       6          6
33       2          2
77       7          4
4        1          1
93       1          1
78       2          2
12       4          4
31       1          1
76       3          4
89       5          1
26       5          1
42       6          6
70       1          1
15       7          6

## Visualizing crossvalidation score

The glass model was able to predict correctly 61% of the time. We can see that for some instances, the model failed to identify the correct class types. Rest everything was identified correctly.